instrumentation/spring/spring-kafka-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/kafka/InstrumentedBatchInterceptor.java

```diff
@@ -0,0 +1,106 @@
+
+package io.opentelemetry.javaagent.instrumentation.spring.kafka;
+
+import static io.opentelemetry.javaagent.instrumentation.spring.kafka.SpringKafkaSingletons.processInstrumenter;
+import static io.opentelemetry.javaagent.instrumentation.spring.kafka.SpringKafkaSingletons.receiveInstrumenter;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.javaagent.instrumentation.kafka.KafkaConsumerIteratorWrapper;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.springframework.kafka.listener.BatchInterceptor;
+
+public final class InstrumentedBatchInterceptor<K, V> implements BatchInterceptor<K, V> {
+  private final ContextStore<ConsumerRecords<K, V>, State<K, V>> contextStore;
+  @Nullable private final BatchInterceptor<K, V> decorated;
+
+  public InstrumentedBatchInterceptor(
+      ContextStore<ConsumerRecords<K, V>, State<K, V>> contextStore,
+      @Nullable BatchInterceptor<K, V> decorated) {
+    this.contextStore = contextStore;
+    this.decorated = decorated;
+  }
+
+  @Override
+  public ConsumerRecords<K, V> intercept(
+      ConsumerRecords<K, V> consumerRecords, Consumer<K, V> consumer) {
+
+    Context parentContext = Context.current();
+
+    // create spans for all records received in a batch
+    List<SpanContext> receiveSpanContexts = traceReceivingRecords(parentContext, consumerRecords);
+
+    // then start a span for processing that links all those receive spans
+    BatchRecords<K, V> batchRecords = BatchRecords.create(consumerRecords, receiveSpanContexts);
+    if (processInstrumenter().shouldStart(parentContext, batchRecords)) {
+      Context context = processInstrumenter().start(parentContext, batchRecords);
+      Scope scope = context.makeCurrent();
+      contextStore.put(consumerRecords, State.create(batchRecords, context, scope));
+    }
+
+    return decorated == null ? consumerRecords : decorated.intercept(consumerRecords, consumer);
+  }
+
+  private List<SpanContext> traceReceivingRecords(
+      Context parentContext, ConsumerRecords<K, V> records) {
+    List<SpanContext> receiveSpanContexts = new ArrayList<>();
+
+    Iterator<ConsumerRecord<K, V>> it = records.iterator();
+    // this will forcefully suppress the kafka-clients CONSUMER instrumentation even though there's
+    // no current CONSUMER span
+    // this instrumentation will create CONSUMER receive spans for each record instead of
+    // kafka-clients
+    if (it instanceof KafkaConsumerIteratorWrapper) {
+      it = ((KafkaConsumerIteratorWrapper<K, V>) it).unwrap();
```

Comment:
It definitely is better to have it disabled by default. For a broader picture, I'm planning to instrument the most commonly used "higher-level" kafka consumer APIs (spring, vertx, reactor, ... anything else?) and after they all have their own separate instrumentations I believe we should have the kafka-clients consumer one disabled by default (which is exactly what @trask suggested in https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1947#issue-771315965). It might take some time though 😅 
For now this hack must suffice, but I'd love to turn off the kafka-clients instrumentation completely sometime.